### PR TITLE
Add unit tests for the controller

### PR DESCRIPTION
**Pull request by vaijab**

### CODE REVIEW

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -39,7 +45,7 @@ type Controller struct {
 // Config represents a controller configuration.
 type Config struct {
 	Cloud    cloudprovider.Interface
-	UserData *userdata.UserData
+	UserData userdata.UserDater
```
Comment (jits) [inline pkg/controller/controller.go]: `UserDater`? Are we branching out into the online dating sector? ;-)

File: pkg/controller/controller_test.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+	m.NodePooler.On("CreateMasterPool", cluster.MasterPool).Return(nil)
+
+	if err := ctrl.CreateCluster(cluster, model.Assets{}); err != nil {
+		t.Error(err)
+	}
+
+	m.Clusters.AssertExpectations(t)
+	m.UserData.AssertExpectations(t)
+	m.NodePooler.AssertExpectations(t)
+}
```
Comment (jits) [inline pkg/controller/controller_test.go]: How about assertions on `m.Provider`?

File: pkg/controller/controller_test.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+	cluster := model.Cluster{
+		Name:       "foo",
+		MasterPool: model.MasterPool{NodePool: makeNodePool("foo", "master")},
+	}
+
+	m.Clusters.On("GetClusters", cluster.Name).Return([]*model.Cluster{&cluster}, nil).Once()
+
+	if err := ctrl.CreateCluster(cluster, model.Assets{}); err == nil {
+		t.Error("should have returned an error")
+	}
```
Comment (jits) [inline pkg/controller/controller_test.go]: There should be a helper in the test framework to catch specific types of errors here?

The problem here is _any_ error returned will cause this to pass, which could mask unwanted error types, so best to check for specific error types, and use a test framework specific way to do it (if possible).

File: pkg/controller/controller_test.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+	p := model.MasterPool{
+		NodePool: makeNodePool(clusterName, "master"),
+	}
+
+	m.Clusters.On("GetClusters", clusterName).Return([]*model.Cluster{&model.Cluster{Name: clusterName}}, nil).Once()
+	m.NodePooler.On("GetMasterPools", clusterName, "").Return([]*model.MasterPool{&p}, nil)
+
+	if err := ctrl.CreateMasterPool(p); err == nil {
+		t.Error("should have received an error")
+	}
```
Comment (jits) [inline pkg/controller/controller_test.go]: Likewise here (re: error checking)

File: pkg/controller/controller_test.go (hunk 2 above)
Comment (vaijab) [inline pkg/controller/controller_test.go]: Yeah, I need to change something to make it work, was gonna do it.

File: pkg/controller/controller.go (hunk 1 above)
Comment (vaijab) [inline pkg/controller/controller.go]: Yep. You found a hidden feature!

File: pkg/controller/controller_test.go (hunk 3 above)
Comment (vaijab) [inline pkg/controller/controller_test.go]: Good point.

**Comment by vaijab** [comment]
@jits updated to address your suggestions. Please take another look.

**Pull request merged**